### PR TITLE
Feature/91 position od

### PR DESCRIPTION
**Pull request by thibaudteil**
* **Tickets addressed:** bsk-091
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description
This tickets adds a SRuKF filter to process full camera position vectors. It is nearly an exact replica of the flyby filter, motivating the generation of an interface.

## Verification
As with other filters, the test checks for propagation and measurements. The post fit residuals are seen to reconverge into the measurement noise when the dynamics shift.

## Documentation
RST documentation

## Future work
Refactor part of the interface

### CODE REVIEW

File: src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.cpp [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+    }
+
+    /*! - Assemble the A matrix for QR decomposition as seen in equation 20 in the reference document*/
+    for (size_t i = 1; i < this->numberSigmaPoints; i++)
+    {
+        A.col(i-1) = sqrt(this->wC(i))*(this->sigmaPoints.col(i) - this->xBar);
+    }
+
+    A.block(0, this->numberSigmaPoints-1, this->state.size(), this->state.size()) =
+            this->cholProcessNoise;
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.cpp]: :large_blue_circle: Could throw this back inline above.

File: src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.cpp [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+        this->yMeas.col(j) = this->sigmaPoints.col(j).head(3);
+    }
+}
+
+/*! Compute the post fit residuals if the measurement data was fresh.
+ * The post fits are y - ybar if a measurement was read, if observations are not present
+ * a flag is raised to not compute post fit residuals
+@return void
+ */
+void PositionODuKF::computePostFitResiudals()
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.cpp]: :large_orange_diamond: typo in `Residuals`

File: src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.cpp [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+    this->measurementModel();
+    /*! - Compute the value for the yBar parameter (equation 23)*/
+    this->postFits.setZero(this->obs.size());
+    Eigen::VectorXd yBar;
+    yBar.setZero(this->obs.size());
+    for(size_t i=0; i<this->numberSigmaPoints; i++){
+        yBar += this->wM(i)*this->yMeas.col(i);
+    }
+    this->postFits = this->obs - yBar;
+    this->computePostFits = true;
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.cpp]: :large_orange_diamond: Are the implied semantics of `computePostFits`, post fits are available/have been computed (given the time tag from the measured data is new compared to previous step), I think that's right? I see the header file comment.

File: src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.h [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    void UpdateState(uint64_t CurrentSimNanos) override;
+
+private:
+    void timeUpdate(const double updateTime);
+    void measurementUpdate();
+    void measurementModel();
+    void readFilterMeasurements();
+    void writeOutputMessages(uint64_t CurrentSimNanos);
+    void computePostFitResiudals();
+    Eigen::MatrixXd qrDecompositionJustR(const Eigen::MatrixXd input) const;
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.h]: :large_orange_diamond: Let's quickly touch base about adjusting some of these to `const` references and removing the const on some parameters if we truly want them copied in to the function.

File: src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.cpp [hunk 5]
```diff
@@ -0,0 +1,509 @@
+/*
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.cpp:1]: :large_blue_circle: It looks like SourceTree is adding some text to the commit messages based on them being cherry picked?

**Review by patkenneally — APPROVED** [review]
Legend:
:small_red_triangle_down: *Issues* to address before merge
:large_orange_diamond: *Requests* that should not block merge, but should at least be discussed
:large_blue_circle: *Recommendations* that can be ignored if desired

File: src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.cpp (hunk 1 above)
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.cpp]: Done

File: src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.cpp (hunk 2 above)
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.cpp]: Done

File: src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.cpp (hunk 3 above)
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.cpp]: changed it to measurementRead to be more consistent, and placed it in the readMeasurement function

File: src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.cpp (hunk 5 above)
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.cpp:1]: Fixed

File: src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.h (hunk 4 above)
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/positionODuKF/positionODuKF.h]: Fixed

**Pull request merged**